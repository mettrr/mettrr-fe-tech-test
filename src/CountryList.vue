<template>
    <div id="app">
        <div class="dropdown">
            <div class="dropdown__title">
                List of Countries
                <i class="dropdown__icon fas fa-chevron-circle-down"></i>
            </div>
            <ul class="dropdown__list">
                <li v-for="(country) in countries"
                class="dropdown__item">
                    {{ country.name }}
                </li>
            </ul>
        </div>

        <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.0.10/css/all.css" integrity="sha384-+d0P83n9kaQMCwj8F4RJB66tzIwOKmrdb46+porD/OvrJ+37WqIM7UoBtwHO6Nlg" crossorigin="anonymous">
    </div>
</template>

<script>
    import axios from 'axios';

    export default {
        name: 'CountryList',
        data() {
            return {
                countries: [],
                // The following no fly list is based on this article, found by literally google searching "country no fly list" https://www.politico.com/interactives/2017/trump-travel-ban-country-list-map/ In a real world scenario, I would request a no fly list or, better yet, an endpoint to a no fly list.
                noFlyList: [
                    'TD',
                    'KP',
                    'VE',
                    'IR',
                    'LY',
                    'SO',
                    'SY',
                    'YE',
                ],
                countriesEndPoint: 'https://api.myjson.com/bins/16i597',
            };
        },
        created() {
            axios.get(this.countriesEndPoint)
                .then((r) => {
                    for (let i = 0; i < r.data.length; i += 1) {
                        if (!this.noFlyList.includes(r.data[i].code)) {
                            this.countries.push(r.data[i]);
                        }
                    }
                })
                .catch((e) => {
                    console.log(e);
                });
        },
    };
</script>

<style lang="scss">
    body {
        font-family: 'Nunito Sans', sans-serif;

        #app {
            margin: 0 auto;
            width: 300px;
        }
    }

    .dropdown {
        box-sizing: border-box;
        padding: .75em;
        background-color: #b4ffaa;
        border: 1px solid;
        position: relative;
        cursor: default;

        &:hover {
            .dropdown {
                &__list {
                    max-height: 300px;
                    overflow: scroll;
                }

                &__icon {
                    transform: rotate(0deg);
                }
            }
        }

        &__title {
            font-weight: bold;
            font-size: 1.25em;
        }

        &__icon {
            float: right;
            transform: rotate(-90deg);
            transition: transform .3s;
        }

        &__list {
            position: absolute;
            top: 48px;
            left: 0;
            box-sizing: border-box;
            max-height: 0;
            list-style: none;
            border: 1px solid;
            overflow: hidden;
            margin: 0;
            padding: 0;
            transition: visibility 0s, max-height .3s ease-in;
        }

        &__item {
            padding: 1em;
            margin: 0 1em;
            border-bottom: 1px solid #cecece;
        }
    }
</style>