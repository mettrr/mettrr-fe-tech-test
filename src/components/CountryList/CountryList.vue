<template>
    <div>
        <h2>Country List</h2>
        <router-link to="/">Home</router-link>
        <div class="dropdown">
            <div class="dropdown-selected-item" 
                v-on:click="toggleDropdownList">
                <p v-if="selectedListItem">{{ selectedListItem.name }}</p>
            </div>
            <ul class="dropdown-list"
                v-bind:class="{active: dropdownListIsActive}">
                <li v-for="(country, index) in countryList" 
                    :key="`fruit=${index}`"
                    v-on:click="setSelectedListItem(country)">
                    {{ country.name }}
                </li>
            </ul>
        </div>
        <h3 v-if="selectedListItem">You selected {{selectedListItem.name}}</h3>
    </div>
</template>

<script>
import getCountryListExceptNoFlyList from '../../data/getCountryListExceptNoFlyList';

export default {
    data() {
        return {
            countryList: [],
            selectedListItem: false,
            dropdownListIsActive: false,
            hasError: false,
            errorMessage: '',
        };
    },
    mounted() {
        this.getCountryItems();
    },
    methods: {
        toggleDropdownList() {
            this.dropdownListIsActive = !this.dropdownListIsActive;
        },
        setSelectedListItem(country) {
            this.selectedListItem = country;
            this.dropdownListIsActive = false;
        },
        getCountryItems() {
            const self = this;

            // Reset feedback indicators
            self.hasError = false;
            self.errorMessage = '';

            getCountryListExceptNoFlyList().then((list) => {
                self.countryList = list;
                [self.selectedListItem = 0] = list;
            }).catch((err) => {
                self.hasError = true;
                self.errorMessage = err.message;
            });
        },
    },
};
</script>

<style lang="scss">
    .dropdown {
        position: relative;
        margin-top: 15px;

        .dropdown-selected-item {
            cursor: pointer;
            box-shadow: 0 2px 8px rgba(0,0,0,.15);
            padding: 1px 0;
            border-radius: 4px;

            p {
                margin: 3px 0;
            }
        }

        .dropdown-list {
            margin: 0;
            padding: 0;
            list-style: none;
            display: none;
            position: absolute;
            top: 130%;
            box-shadow: 0 2px 8px rgba(0,0,0,.15);
            border-radius: 4px;
            background-color: #fff;

            &.active {
                display: block;
                max-height: 200px;
                overflow-y: scroll;
            }

            &::-webkit-scrollbar {
                width: 6px;
                background-color: #F5F5F5;
            }

            &::-webkit-scrollbar-thumb {
                background-color: #000000;
            }

            &::-webkit-scrollbar-track {
                -webkit-box-shadow: inset 0 0 6px rgba(0,0,0,0.3);
                background-color: #F5F5F5;
            }

            li {
                padding: 6px 0;
                cursor: pointer;

                &:hover {
                    background-color: #F5F5F5;
                }
            }
        }   
    }
</style>