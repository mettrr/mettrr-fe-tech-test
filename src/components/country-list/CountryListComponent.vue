<template>
    <div>
      <label>Select a country</label>
      <br><br>
      <select v-model="selectedCountry" class="form-control">
          <option selected disabled value="disabled">Country...</option>
          <option v-for="country in countries" v-bind:value="country.code" v-bind:key="country.code">
            {{ country.name }} ({{Â country.code }})
          </option>
        </select>
    </div> 
</template>

<script>
import axios from 'axios';

export default {
    name: 'CountryListComponent',
    data() {
        return {
            countries: [],
            noFlyList: ['VE', 'SY', 'IR', 'SO', 'KR', 'YE'], // Those are the countries banned from the US
            selectedCountry: '',
            errors: [],
        };
    },
    methods: {
        getCountries() {
            const that = this;
            const apiUrl = 'https://api.myjson.com/bins/16i597';
            axios.get(apiUrl) // Getting values from the API with a GET method
                .then(response => {
                    console.log('response open', response);
                    response.data.forEach(element => {
                        if (!this.noFlyList.includes(element.code)) { // Needed to exclude the countries banned from the US
                            that.countries.push(element);
                        }
                    });
                })
                .catch(e => {
                    this.errors.push(e);
                });
        },
    },
    created() {
        this.getCountries(); // Getting counties from the API
    },
};
</script>

<style lang="scss">
    select {
        border: 1px solid #7550ff;
        height: 30px;
        outline: none;
    }
</style>