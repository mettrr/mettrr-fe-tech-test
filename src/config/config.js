export const apiEndPoint = 'https://api.myjson.com/bins/16i597';
export const NOFLYLIST = [
    'TD', 'KP', 'VE', 'IR',
    'LY', 'SO', 'SY', 'YE',
];
