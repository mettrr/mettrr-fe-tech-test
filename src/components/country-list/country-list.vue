<template>
    <div class='countryList__Container'>
        <div class='collapse' v-on:click="dropDown">
            <p>{{ title }}</p>
            <img :class="[collapse ? '--collapse' : '']"
                class='caret__Image' src='../../assets/arrow-64.png' />
        </div>
        <p class='error__Msg' 
            v-if="fetchError">Oops! Error in Fetching country data...</p>
            <ul class='countryList'
                v-if="!fetchError"
                :class="[collapse ? '--collapse-list' : '']">
                <li :class="[collapse ? '--animate' : '']"
                v-for="country in countries" 
                :key="country.code">{{country.name}}</li>
            </ul>
    </div>
    
</template>
<script>

import { getCountry } from './country-list.api';
import { removeNoFly } from './country-list.funcs';

export default {
    name: 'countryList__Container',
    data() {
        return {
            countries: [],
            title: 'Click Me',
            collapse: false,
            fetchError: false,
        };
    },
    created() {
        getCountry()
            .then(countries => {
                this.countries = removeNoFly(countries);
            })
            .catch(error => {
                this.fetchError = true;
            });
    },
    methods: {
        dropDown() {
            this.collapse = !this.collapse;
        },
    },

};
</script>
<style lang="scss">
$border: 1px solid lighten(grey, 40%);

.collapse {
    display: flex;
    flex-flow: row wrap;
    justify-content: space-between;
    align-items: center;
    padding: 0 0.5rem;
    border: $border;
    color: gray;
    font-weight: 400;
}
.caret__Image {
    height: 20px;
    width: 20px;
    transform: rotate(0deg);
    transition: 0.5s all ease;

    &.--collapse {
        transform: rotate(90deg);
    }
}

ul {
    list-style-type: none; 
    margin: 0;
    padding: 0;
    background: lighten(blue, 50%);
    overflow-y: scroll;
    border: $border;
    height: 0px;
    transition: 0.5s all ease-out;

    &.--collapse-list {
        height: 400px;
        display: block;
    }
}
ul li {
    text-align: left;
    margin: 0.1rem 0;
    padding: 0.7rem;
    font-size: 1rem;
    color: grey;
    font-weight: 200;
    letter-spacing: 0.5px;
    transform: translateY(-40px);
    transition: 0.5s all ease-in;
    border-left: 3px solid grey;
    border-bottom: 1px solid grey;
    &:hover {
        background: linear-gradient(45deg,lighten(grey, 10%), lighten(grey, 60%));  
        color: white; 
        font-weight: 400;
    }
    &.--animate {
        transform: translateY(0);
    }
}
.error__Msg {
    color: lighten(red, 30%);
    font-weight: 400;
    font-size: 1.5rem;
}
</style>