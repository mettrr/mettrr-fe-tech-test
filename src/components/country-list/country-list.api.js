import axios from 'axios';
import { apiEndPoint } from '../../config/config';

export const getCountry = () => {
    const promise = new Promise((resolve, reject) => {
        axios
            .get(`${apiEndPoint}`)
            .then(countries => resolve(countries.data))
            .catch(error => reject(error));
    });
    return promise;
};
