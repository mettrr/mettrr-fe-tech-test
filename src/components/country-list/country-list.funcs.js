import { NOFLYLIST } from '../../config';

export const removeNoFly = (countries) => {
    const newCountries = countries.filter(country => !NOFLYLIST.includes(country.code));
    return newCountries;
};
