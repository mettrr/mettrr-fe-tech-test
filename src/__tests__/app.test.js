import * as vTests from 'vue-test-utils';
import moxios from 'moxios';
import { mount } from 'avoriaz'
import { countryRoute } from '../config';
import { getCountry } from '../components/country-list/country-list.api';
import CountryList from '../components/country-list/country-list.vue';
import { createRenderer } from 'vue-server-renderer';
import axios from 'axios';

describe('CountryList ', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = vTests.mount(CountryList);
        moxios.install(axios);
    });
    afterEach(() => {
        moxios.uninstall(axios);
    })
    it('renders', () => {
        const renderer = createRenderer()

        renderer.renderToString(wrapper.vm, (err, str) => {
            if (err) throw new Error(err)
            expect(str).toMatchSnapshot()
        })
    })

    it('dropwdown function is called', () => {
        const saveMock = jest.fn()
        wrapper.vm.dropDown = saveMock
        wrapper.update()

        wrapper.find('.collapse').trigger('click')
        expect(saveMock.mock.calls.length).toBe(1)
    })

    it('valid api route', async () => {
        moxios.stubRequest(countryRoute, {
            status: 200,
            response: [
                {
                    "name": "Afghanistan",
                    "code": "AF"
                },
                {
                    "name": "Ã…land Islands",
                    "code": "AX"
                },
                {
                    "name": "Albania",
                    "code": "AL"
                },
                {
                    "name": "Algeria",
                    "code": "DZ"
                },
                {
                    "name": "American Samoa",
                    "code": "AS"
                },
            ]
        })
        let wrapper = mount(CountryList)

        const result = await getCountry()
        expect(result.length).toEqual(wrapper.data().countries.length)
    })
})