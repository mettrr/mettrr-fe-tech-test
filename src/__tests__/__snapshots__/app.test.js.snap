// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryList  renders 1`] = `"<div data-server-rendered=\\"true\\" class=\\"countryList__Container\\"><div class=\\"collapse\\"><p>Click Me</p> <img src=\\"../../assets/arrow-64.png\\" class=\\"caret__Image\\"></div> <!----> <ul class=\\"countryList\\"></ul></div>"`;
